core/src/main/java/org/apache/accumulo/core/lock/ServiceLockPaths.java

```diff
@@ -0,0 +1,385 @@
+      if (this == obj) {
+        return true;
+      }
+      if (obj == null) {
+        return false;
+      }
+      if (getClass() != obj.getClass()) {
+        return false;
+      }
+      ServiceLockPath other = (ServiceLockPath) obj;
+      if (path == null) {
+        if (other.path != null) {
+          return false;
+        }
+      } else if (!path.equals(other.path)) {
+        return false;
+      }
+      if (resourceGroup == null) {
+        if (other.resourceGroup != null) {
+          return false;
+        }
+      } else if (!resourceGroup.equals(other.resourceGroup)) {
+        return false;
+      }
+      if (server == null) {
+        if (other.server != null) {
+          return false;
+        }
+      } else if (!server.equals(other.server)) {
+        return false;
+      }
+      if (type == null) {
+        if (other.type != null) {
+          return false;
+        }
+      } else if (!type.equals(other.type)) {
+        return false;
+      }
+      return true;
+    }
+
+    @Override
+    public int hashCode() {
+      final int prime = 31;
+      int result = 1;
+      result = prime * result + ((path == null) ? 0 : path.hashCode());
+      result = prime * result + ((resourceGroup == null) ? 0 : resourceGroup.hashCode());
+      result = prime * result + ((server == null) ? 0 : server.hashCode());
+      result = prime * result + ((type == null) ? 0 : type.hashCode());
+      return result;
+    }
+
+    @Override
+    public String toString() {
+      return this.path;
+    }
+
+  }
+
+  private final ClientContext ctx;
+
+  public ServiceLockPaths(ClientContext context) {
+    this.ctx = context;
+  }
+
+  private static String determineServerType(final String path) {
+    if (path.contains(Constants.ZGC_LOCK)) {
+      return Constants.ZGC_LOCK;
+    } else if (path.contains(Constants.ZMANAGER_LOCK)) {
+      return Constants.ZMANAGER_LOCK;
+    } else if (path.contains(Constants.ZMONITOR_LOCK)) {
+      return Constants.ZMONITOR_LOCK;
+    } else if (path.contains(Constants.ZMINI_LOCK)) {
+      return Constants.ZMINI_LOCK;
+    } else if (path.contains(Constants.ZCOMPACTORS)) {
+      return Constants.ZCOMPACTORS;
+    } else if (path.contains(Constants.ZSSERVERS)) {
+      return Constants.ZSSERVERS;
+    } else if (path.contains(Constants.ZDEADTSERVERS)) {
+      // This has to be before TSERVERS
+      return Constants.ZDEADTSERVERS;
+    } else if (path.contains(Constants.ZTSERVERS)) {
+      return Constants.ZTSERVERS;
+    } else {
+      throw new IllegalArgumentException("Unhandled to determine server type from path: " + path);
+    }
+  }
+
+  /**
+   * Parse a ZooKeeper path string and return a ServiceLockPath
+   */
+  public static ServiceLockPath parse(Optional<String> serverType, String path) {
+    Objects.requireNonNull(serverType);
+    Objects.requireNonNull(path);
+
+    final String type = serverType.isEmpty() ? determineServerType(path) : serverType.orElseThrow();
```

Comment:
```suggestion
    final String type = serverType.orElseGet(()->determineServerType(path));
```

Reply:
Nice, I didn't know about that. Updated in 69f40d6.